middleware/timeout/timeout.go

```diff
@@ -21,42 +22,34 @@ func New(h fiber.Handler, config ...Config) fiber.Handler {
 
 		timeout := cfg.Timeout
 		if timeout <= 0 {
-			return runHandler(ctx, h, cfg)
+			return h(ctx)
 		}
 
-		parent := ctx.Context()
-		tCtx, cancel := context.WithTimeout(parent, timeout)
-		ctx.SetContext(tCtx)
-		defer func() {
-			cancel()
-			ctx.SetContext(parent)
+		err := make(chan error, 1)
+		go func() {
+			err <- h(ctx)
 		}()
-
-		err := runHandler(ctx, h, cfg)
-
-		if errors.Is(tCtx.Err(), context.DeadlineExceeded) && err == nil {
+		select {
+		case err := <-err:
+			if err != nil && (len(cfg.Errors) > 0 && isCustomError(err, cfg.Errors)) {
+				if cfg.OnTimeout != nil {
+					if toErr := cfg.OnTimeout(ctx); toErr != nil {
+						return toErr
+					}
+				}
+				return fiber.ErrRequestTimeout
+			}
+			return err
+		case <-time.After(timeout):
 			if cfg.OnTimeout != nil {
-				return cfg.OnTimeout(ctx)
+				err := cfg.OnTimeout(ctx)
+				ctx.RequestCtx().TimeoutErrorWithResponse(&ctx.RequestCtx().Response)
```

Comment:
This is a problem, you have to create a brand new response object not the one from the ctx since it gets recycle.

Reply:
The function TimeoutErrorWithResponse already creates a new response inside